haskell-style.md

```diff
@@ -660,3 +660,329 @@ module Driver
     ) where
 
   ```
+
+## Haddocks
+
+Haskell modules should be commented with valid Haddock documentation. We are not
+yet requiring a certain level of coverage, but it is strongly encouraged.
+
+### General guides
+
+- Use proper Haddock [markup][]
+- Link all identifiers, anywhere they appear
+
+  Bad
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @Right FlipFlop@, otherwise a @Left@
+  --
+  ```
+
+  Good
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @'Right' 'FlipFlop'@, otherwise a
+  -- @'Left'@.
+  --
+  ```
+
+- Use leading documentation (`-- |`) for top-level definitions and trailing
+  documentation (`-- ^`) for record attributes and function arguments
+
+  Bad
+
+  ```hs
+  data Foo = Foo
+    {
+    -- | Foo's foo
+      fooFoo :: Foo
+    -- | Foo's bar
+    , fooBar :: Bar
+    }
+  -- ^ What even is a Foo?
+  ```
+
+  Good
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoo :: Foo
+    -- ^ Foo's foo
+    , fooBar :: Bar
+    -- ^ Foo's bar
+    }
+  ```
+
+  Best, for this case
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoo :: Foo -- ^ Foo's foo
+    , fooBar :: Bar -- ^ Foo's bar
+    }
+  ```
+
+  **NOTE**: do not align trailing documentation in context-sensitive ways.
+
+  Bad
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoos :: [Foo] -- ^ Foo's foos
+    , fooBar :: Bar    -- ^ Foo's bar
+    }
+  ```
+
+  Good
+
+  ```hs
+  data Foo = Foo
+    { fooFoos :: [Foo] -- ^ Foo's foos
+    , fooBar :: Bar -- ^ Foo's bar
+    }
+  ```
+
+### Summaries
+
+Summaries must be a single, short (e.g. non-wrapping), capitalized sentence; not
+punctuated, and in a declarative tense.
+
+A Summary should complete the sentence:
+
+> This (module|type|attribute|function|argument)... {Summary}
+
+Most rules that would apply to [commit messages][tbaggery] apply here.
+
+Bad
+
+```hs
+-- | Be careful here, this is tricky!
+
+-- | Here we're returning the Admins that can access the other thing by virtue
+-- of the fact that they are this thing
+
+-- | Does a thing. Is partial because of random reason
+```
+
+Good
+
+```hs
+-- | Represents a value that may or may not be present
+{- This type represents a value... -}
```

Comment:
The `{- ... -}` comments look like they're meant to represent the comment body. Is this supposed to be showing how the summary completes the "This <thing> ..." sentence?

Reply:
> Is this supposed to be showing how the summary completes...

That was the intention, I can remove it if it's confusing.

Reply:
Yes please, much confuse.